Updating discovery docs (#10859)
solo-io/gloo: docs/content/installation/advanced_configuration/fds_mode.md

---
title: Configuring Discovery
weight: 10
description: Set up automatic discovery for upstreams and functions. automatic function discovery.
---

> Gloo Gateway supports automatically discovering upstreams and functions. 

Author: Added UDS to the FDS guide so the discovery info is together in one place.

```bash
kubectl edit -n gloo-system settings.gloo.solo.io
```

Example Settings CR:
```yaml
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: gloo.solo.io/v1
kind: Settings
metadata:
  labels:
    app: gloo
    gloo: settings
  name: default
  namespace: gloo-system
spec:
  bindAddr: 0.0.0.0:9977
  discoveryNamespace: gloo-system
  kubernetesArtifactSource: {}
  kubernetesConfigSource: {}
  kubernetesSecretSource: {}
  refreshRate: 60s
  # add the following lines
  discovery:
    enabled: true
```
{{% /tab %}} 
{{< /tabs >}}

### More information about UDS {#more-info}

For more information, check out the following guides:

* [Discovered Upstreams]({{< versioned_link_path fromRoot="/guides/traffic_management/destination_types/discovered_upstream/" >}})
* [Configuration via Annotations]({{< versioned_link_path fromRoot="/guides/traffic_management/destination_types/discovered_upstream/discovered-upstream-configuration/" >}})

---

> ## Function Discovery Service (FDS) {#fds}

Author: Reorg'd the FDS content a bit

{{< /highlight >}}

### How FDS works

Using FDS means that the Gloo Gateway `discovery` component will make HTTP requests to all `Upstreams` known to Gloo Gateway trying to discover functions. This behavior causes increased network traffic and may be undesirable if it causes unexpected behavior or logs to appear in the services Gloo Gateway is attempting to poll. For this reason, we may want to restrict the manner in which FDS polls services.

Gloo Gateway allows whitelisting/blacklisting services, either by namespace or on the individual service level.

We can use these configuration settings to restrict FDS to discover only the namespaces or individual services we choose.

### Enable FDS mode

> You can enable FDS mode with Helm or the Settings CR in the `gloo-system` namespace.

Reviewer: same thing, maybe mention it is not enabled by default and needs to explicitly be turned on

Author: I have this note at the top of the page. Or do you think I should also reiterate it in both the H2s for UDS/FDS?

```
{{% notice note %}}
Because discovery is a resource-intensive process, both discovery services are disabled by default.
{{% /notice %}}
```

    # WHITELIST is the default value
    fdsMode: WHITELIST
    fdsOptions:
      # Enterprise-only: set to false to disable automated GraphQL schema generation as part of FDS.
      # true is the default value (enabled)
      graphqlEnabled: true
```
{{% /tab %}} 
{{< /tabs >}}

### FDS modes

> FDS can run in one of 3 modes:

Reviewer: ```suggestion
FDS can run in one of three modes:
```

or can do `1 of 3`

> * `BLACKLIST`: The most liberal FDS polling policy. Using this mode, FDS will poll any service unless its namespace or the service itself is explicitly blacklisted.
> * `WHITELIST`: A more restricted FDS polling policy. Using this mode, FDS will poll only those services who either live in an explicitly whitelisted namespace, or themselves are are explicitly whitelisted. *`WHITELIST` is the default mode for FDS*.
> * `DISABLED`: FDS will not run. **Upstream Discovery Service** (UDS) will still run as normal.

Reviewer: i think this was existing content, but just while we're here. feel free to change
```suggestion
* `BLACKLIST`: The most liberal FDS polling policy. FDS polls all services, unless the service or the service's namespace is blacklisted.
* `WHITELIST` (default mode): A more restrictive FDS polling policy. FDS polls only services that are either whitelisted or exist in a whitelisted namespace.
* `DISABLED`: FDS does not run. UDS continues to run as normal.
```
